Feature/seq editor cleanup

cohansen opened the pull request:
Addresses all of the `IN PROGRESS` tasks in https://github.com/NASA-AMMOS/aerie-ui/issues/1291.

cohansen committed:
Cleanup NumEditor and convert value to a number

cohansen committed:
Added tooltip support for repeat args

cohansen committed:
Removed utils.ts again

cohansen committed:
Cleaned up DictionaryTable so theres only one selection type now

cohansen committed:
Cleaned up dictionary permissions

cohansen committed:
Removed aerie-phoenix-wordmark

cohansen committed:
Disable seqn and seqjson exports when there is no sequence

cohansen committed:
Added styles to the selected command section

cohansen committed:
Converted fswCommandArgDefault to use a switch

cohansen committed:
Renamed parcel to parameter dictionaries

cohansen committed:
Some more cleanup

AaronPlave commented on src/utilities/permissions.ts:
```diff
@@ -513,8 +513,8 @@ const queryPermissions = {
   },
-  DELETE_PARCEL_TO_PARAMETER_DICTIONARIES: (user: User | null): boolean => {
-    return isUserAdmin(user) || getPermission([Queries.DELETE_PARCEL_TO_PARAMETER_DICTIONARY], user);
+  DELETE_PARCEL_TO_DICTIONARY_ASSOCIATION: (user: User | null): boolean => {
```
I'm seeing that I no longer have permission to delete anything from the tables within the parcels page whereas on develop I can. Is this intended?
<img width="375" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/da7180fc-2607-43cf-8800-3e92c0087607">

cohansen replied:
Which role are you logged in as? This should be admin only.

AaronPlave replied:
This is as admin role

AaronPlave commented on src/components/parcels/ParcelForm.svelte:
Seeing inconsistent behavior regarding:
> Able to uncheck a dictionary (and not select a new one) from a parcel, but saving doesn't persist. - `IN PROGRESS`

https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/94637747-0f4e-4d47-bb02-d7c728fd9c08

AaronPlave replied:
Still seeing this issue

AaronPlave replied:
https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/0021dd97-242e-4626-8bef-ed06de6e26ad

AaronPlave commented on src/components/parcels/DictionaryTable.svelte:
In multiselect mode it would be nice if clicking on an already selected row deselected the row to mirror what happens in single select mode

AaronPlave replied:
Would there be some way to remove the row highlight as well? Kind of confusing that it remains.

AaronPlave replied:
Punting on this

AaronPlave commented on src/components/sequencing/form/SelectedCommand.svelte:
```diff
@@ -191,21 +193,29 @@
+          </fieldset>
+        {/if}
+
+        <fieldset>
```
Could you use the `Collapse` component here? I can also take a crack at styling this if that'd be helpful.

AaronPlave replied:
Done

AaronPlave commented on src/components/sequencing/SequenceEditor.svelte:
Seeing an issue with the tooltip in the editor when you hover over unknown args
<img width="1390" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/5e2d0571-28f2-4ed5-80da-3d8ce8f4401f">

AaronPlave replied:
This also breaks the Selected Command editor - can't edit any inputs or remove unknown arguments

AaronPlave commented on src/components/sequencing/SequenceEditor.svelte:
Editing an arg value from the Selected Command editor most often does not result in updates to the actual sequence. Can repro using a basic `C SEQ_ECHO "some string"`

cohansen replied:
I found that I had to blur the input to get the changes to propagate, are you seeing this on other types as well?

AaronPlave replied:
This was happening on blur and enter

cohansen committed:
Fixed toggling by clicking a row on the parcel page

AaronPlave commented:
Missing this item from the in progress list:
>Empty state for third column in sequence editor

cohansen committed:
Fixed an issue where changing selected parameter dictionaries for a parcel wasn't enabling the save button

cohansen commented:
> Missing this item from the in progress list:
> 
> > Empty state for third column in sequence editor

I added this empty state to the selected command editor, unless this isn't what you were referencing?

![Screenshot 2024-06-20 at 6 45 57 AM](https://github.com/NASA-AMMOS/aerie-ui/assets/3433719/0174bdf9-7464-4868-ab2e-353df405fa3d)

cohansen committed:
Fixed an issue where selecting remove in the command editor panel didn't remove the broken argument

AaronPlave commented:
> > Missing this item from the in progress list:
> > > Empty state for third column in sequence editor
> 
> I added this empty state to the selected command editor, unless this isn't what you were referencing?
> 
> ![Screenshot 2024-06-20 at 6 45 57 AM](https://private-user-images.githubusercontent.com/3433719/341496429-0174bdf9-7464-4868-ab2e-353df405fa3d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5MDI0OTYsIm5iZiI6MTcxODkwMjE5NiwicGF0aCI6Ii8zNDMzNzE5LzM0MTQ5NjQyOS0wMTc0YmRmOS03NDY0LTQ4NjgtYWIyZS0zNTNkZjQwNWZhM2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjBUMTY0OTU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjQ2ZTdlMTg3MDk2YTQ4NzY2OWE0MjA2MWQ4MGY1YTNhYWVlNzZkZjJjNWNlMWYyOGRiYWU1ZTRkZGZmMjRlNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ds65DX_F025PvknB9FD2nfqu9HN9fwZBQ7Vn5p985Mk)

Ah I see, have to select a parcel first, thanks. Could you add `st-typography-body`?

cohansen committed:
Added st-typography-body to the no selected command text

cohansen committed:
Fixed an issue where the delete button was showing on dictionaries on the parcel page

cohansen committed:
Added some hint text to the Selected Command panel when there isn't a parcel selected

cohansen committed:
Fixed an issue with the selected command panel wasn't propagating changes

cohansen committed:
Removed the shared dictionary permission for creation

cohansen committed:
Changed permissions for the dictionary create button to check every upload permission

cohansen committed:
Fixed one more permission issue and added a div to support selected command testing

cohansen committed:
Added back in the missing seq adaptation name field

duranb commented on src/utilities/new-sequence-editor/sequence-tooltip.ts:
```diff
@@ -35,7 +35,7 @@
 function getTokenPositionInLine(view: EditorView, pos: number) {
   const { from, to, text } = view.state.doc.lineAt(pos);
-  const tokenRegex = /[a-zA-Z0-9_".-]/;
+  const tokenRegex = /[a-zA-Z0-9_".-[\]]/;
```
@cohansen do you mind fixing this warning too?

cohansen committed:
Removed unnecessary part of a regex

duranb reviewed (changes requested):
![Screenshot 2024-06-27 at 12 50 55 PM](https://github.com/NASA-AMMOS/aerie-ui/assets/5290214/70953a84-1293-4cac-90d4-893a23c647b4)

A little outside the scope of this (but became apparent now that it default selects the first command), can we style this button to match the same aesthetic as the other buttons? I think applying the `st-button` class will get you there if not close to it. Also adding some margins to line it up with the above content

duranb reviewed (changes requested):
There might be a weird styling issue here.
![Screenshot 2024-06-27 at 1 03 15 PM](https://github.com/NASA-AMMOS/aerie-ui/assets/5290214/c5863355-ea6d-434d-9623-c2bd19e6b4e4)

I also noticed that selected a command in the preview doesn't populate the `SelectedCommand` component. Should that component be on this page?

duranb commented on src/routes/dictionaries/+page.svelte:
```diff
@@ -35,7 +35,11 @@
+    featurePermissions.commandDictionary.canCreate(data.user) &&
+    featurePermissions.parameterDictionary.canCreate(data.user) &&
+    featurePermissions.sequenceAdaptation.canCreate(data.user);
   $: createButtonDisabled = !files || files?.length === 0 || (isSequenceAdaptation && sequenceAdaptationName === '');
```
it seems like if you create a dictionary, the button still doesn't get disabled despite the input being reset.

cohansen committed:
Fixed some styling issues with the selected command panel

cohansen committed:
Fixed an issue where the upload file button wasn't being disabled after
upload

duranb commented:
I'm also getting a subscription error on `/parcels/new`

```{
    "extensions": {
        "code": "validation-failed",
        "path": "$.selectionSet.parcel_to_parameter_dictionary.args.where.parcel_id._eq"
    },
    "message": "unexpected null value for type 'Int'"
}
```

duranb committed:
allow for panel body to show up when SeqJson editor is "hidden"

cohansen committed:
Cleaned up the sequence store and removed all the parcel and parcel id weirdness

cohansen commented:
@AaronPlave and @duranb this is ready for a re-review.

cohansen committed:
Fixed an issue when creating a new parcel is another user adds one it reset changes

cohansen committed:
Fixed an issue where parcels never became saveable

AaronPlave commented on src/routes/dictionaries/+page.svelte:
Not strictly something from this PR but more noticeable now with sequence adaptations – the create button is not disabled during creation meaning that subsequent clicks on the button cause duplicate creation calls which would be good to avoid.

cohansen replied:
This is ready for re-test.

AaronPlave commented on src/components/parcels/ParcelForm.svelte:
Seeing an issue where I'm able to hit the save button even though I haven't entered a name. The button accepts the click but no network request or error happens.

AaronPlave replied:
https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/da34f1c0-9d1c-4d1a-b53e-c856e7a2bbf1

cohansen replied:
This is fixed and ready for re-test!

AaronPlave committed:
Selected Command styling, fixes, range display

AaronPlave committed:
Range display tweak

AaronPlave commented:
Tweaked the Selected Command editor:
- Use `Collapse` instead of `detail`
- Styling updates
- Display numerical range in collapse header if available
- Bug fix min/max enforcement in numerical input and change to `type="number"`

AaronPlave committed:
NumEditor input validation fix

duranb reviewed (commented):
Uploading a dictionary that already exists with the same name doesn't create a new one even though it says the upload was successful. Is there a way to make it show an error that the upload didn't work?

duranb commented on src/components/sequencing/form/ArgTitle.svelte:
```diff
@@ -41,15 +43,25 @@
+
     if ('units' in argDef) {
-      return `${base} - (${argDef.units})`;
+      return `${base} – (${argDef.units})`;
```
was this intentionally updated to be the longer dash?

AaronPlave replied:
Yep

duranb commented on src/components/sequencing/form/ExtraArgumentEditor.svelte:
```diff
@@ -6,7 +6,7 @@
-<div>
-  <button on:click={setInEditor}>Remove</button>
-  <span>{initVal}</span>
+<div style="display: flex; gap: 8px">
```
Since this is a static style, can this be moved into `<style></style>` tags, please?

duranb commented on src/components/sequencing/form/NumEditor.svelte:
```diff
@@ -7,59 +7,37 @@
-  <!-- on:input={customValidate} -->
-  {#if typeof min === 'number' && typeof max === 'number' && min === max && valFloat !== max}
-    <button on:click={() => setInEditor(max.toString())} title="Set to allowed value">{max}</button>
+  <input class="st-input w-100" type="number" bind:value required {min} {max} step="any" />
```
If the max value of the field is a decimal, pressing the up button doesn't bring the value up to that max, but pressing the down button does bring the value down if the min value contains a decimal.

AaronPlave replied:
Interesting. I needed to use `any` for `step` so that html input validity worked for floats (without a lot of fussing, see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/step for the specifics) but it looks like setting `step` to `any` means that the browser still uses a step of 1 by default. Looks like the browser won't do less than a step to hit the max, oddly. We could maybe capture the specific up and down arrow behaviors on these inputs and perform the stepping ourselves though this feels like an edge case that may not be worth solving given that you can just input the number yourself?

duranb replied:
Ah so it's a browser specific quirk? If so then I agree, it's probably not worth trying to fix the browser behavior.

AaronPlave replied:
Yeah I think this is just how the native behavior works, imagine it'd be the same across all browsers. Figure if it ends up being an issue we could put more time into a better solution

duranb commented on src/components/parcels/ParcelForm.svelte:
```diff
@@ -184,7 +178,7 @@
     if (saveButtonEnabled) {
```
Not relevant to this part of the code, but more for the variable `saveButtonEnabled`.  Small UX bug: if the parcel is saved, when the button text is "Saved", it is still clickable which results in network calls to update.

cohansen replied:
I'm not seeing this behavior locally? I tried clicking it when no changes have been made and also after making some changes and then reverting them.

duranb replied:
Hmm...yeah the exact behavior I initially described is no longer reproducible for me, but I'm seeing that if I rename the parcel and click "Save", it stays as "Save" after the success toast pops up. Are you able to see that too?

cohansen replied:
Yeah, will push up a fix for this!

AaronPlave committed:
Refactor

cohansen commented:
> Uploading a dictionary that already exists with the same name doesn't create a new one even though it says the upload was successful. Is there a way to make it show an error that the upload didn't work?

It actually overwrites the existing dictionary, this is a behavior we've had since before parcels.

cohansen committed:
Feature/remove sequence lang specifics (#1372)

* Moved to a more general seq editor and adaptation

* Fixed an issue where you could save a parcel without a name, fixed being able to upload multiple dictionaries at a time

* Fix plan metadata name input reset on snapshot preview (#1358)

* fix table action icons staying disabled when no longer in snapshot preview
* make table components less plan specific

* Add plan JSON export (#1357)

* add plan JSON export

* Hide plan export button (#1374)

* Release v2.15.0 (#1377)

Co-authored-by: joswig <Joseph.C.Joswig@jpl.nasa.gov>

* Cleanup NumEditor and convert value to a number

Added tooltip support for repeat args

Removed utils.ts again

Cleaned up DictionaryTable so theres only one selection type now

Cleaned up dictionary permissions

Removed aerie-phoenix-wordmark

Disable seqn and seqjson exports when there is no sequence

Added styles to the selected command section

Converted fswCommandArgDefault to use a switch

Renamed parcel to parameter dictionaries

Some more cleanup

Fixed toggling by clicking a row on the parcel page

Fixed an issue where changing selected parameter dictionaries for a parcel wasn't enabling the save button

Fixed an issue where selecting remove in the command editor panel didn't remove the broken argument

Added st-typography-body to the no selected command text

Fixed an issue where the delete button was showing on dictionaries on the parcel page

Added some hint text to the Selected Command panel when there isn't a parcel selected

Fixed an issue with the selected command panel wasn't propagating changes

Removed the shared dictionary permission for creation

Changed permissions for the dictionary create button to check every upload permission

Fixed one more permission issue and added a div to support selected command testing

Added back in the missing seq adaptation name field

Removed unnecessary part of a regex

Fixed some styling issues with the selected command panel

Fixed an issue where the upload file button wasn't being disabled after
upload

allow for panel body to show up when SeqJson editor is "hidden"

Cleaned up the sequence store and removed all the parcel and parcel id weirdness

Fixed an issue when creating a new parcel is another user adds one it reset changes

Fixed an issue where parcels never became saveable

Moved to a more general seq editor and adaptation

Added more generic calling of to output and from input, tried to remove most of the referencs to seqjson

Renamed new-sequence-editor dir to sequence-editor

Fixed a whole bunch of broken imports

Added a helper function to get sequence adaptation globals

---------

Co-authored-by: Bryan <bryan.duran@jpl.nasa.gov>
Co-authored-by: Chet Joswig <cjoswig@yahoo.com>
Co-authored-by: joswig <Joseph.C.Joswig@jpl.nasa.gov>

cohansen committed:
Fixed an issue where renaming the parcel and then saving wasn't setting parcelModified back to false

cohansen committed:
Fixed some tests I broke

cohansen committed:
Fixed all the to and from seq json unit tests

cohansen committed:
Fixed an issue where the save button was flickering on the parcel page

cohansen committed:
Removed unused var

duranb reviewed (approved):
Thanks for the hard work!

AaronPlave reviewed (approved):
LGTM!

cohansen committed:
Added a max-width for the sequence editor tooltips
